File: packages/control-api/src/services/tenant-provisioner.ts
import { db } from './db.js';
import { tenants, auditLogs, users, settings, menuCategories } from '@sibangku/db';
import {
  generateTenantId,
  generateTemporaryPassword,
  generateTenantSlug,
  generatePackageId,
  generateDatabaseIdentifier,
} from '@sibangku/shared';
import pg from 'pg';
import { drizzle } from 'drizzle-orm/node-postgres';
import { migrate } from 'drizzle-orm/node-postgres/migrator';
import path from 'path';
import { fileURLToPath } from 'url';
import bcrypt from 'bcrypt';

const { Client, Pool } = pg;

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Control Database details to connect and create new databases
const controlDbUrl =
  process.env.CONTROL_DATABASE_URL ||
  'postgresql://sibangku:sibangku_dev@localhost:5432/sibangku_control';

export interface ProvisionTenantParams {
  tenantName: string;
  restaurantName: string;
  adminEmail: string;
  trialDays?: number;
}

export interface ProvisionResult {
  tenantId: string;
  tenantCode: string;
  adminEmail: string;
  temporaryPassword: string;
  databaseName: string;
  status: string;
}

export async function provisionTenant(params: ProvisionTenantParams): Promise<ProvisionResult> {
  const { tenantName, restaurantName, adminEmail, trialDays = 60 } = params;

  // 1. Generate unique identifiers
  const tenantId = generateTenantId();
  const tenantSlug = generateTenantSlug(tenantName);
  const tenantCode = tenantSlug.toUpperCase();
  const dbName = generateDatabaseIdentifier(tenantSlug);
  const packageId = generatePackageId(tenantSlug);
  const webIdentifier = `${tenantSlug}.sibangku.example`;
  const apkIdentifier = packageId;
  const brandingIdentifier = `branding_${tenantSlug}`;
  const temporaryPassword = generateTemporaryPassword();

  console.info(`[Provisioner] Starting provisioning for ${tenantName} (ID: ${tenantId}, DB: ${dbName})`);

  // 2. Create the physical database in PostgreSQL
  // Connect to postgres default DB to run CREATE DATABASE
  const parsedUrl = new URL(controlDbUrl);
  // Re-build connection URL pointing to 'postgres' system database
  const systemDbUrl = `${parsedUrl.protocol}//${parsedUrl.username}:${parsedUrl.password}@${parsedUrl.host}/postgres`;

  const pgClient = new Client({ connectionString: systemDbUrl });
  await pgClient.connect();

  try {
    // Check if database already exists
    const dbCheck = await pgClient.query(
      "SELECT 1 FROM pg_database WHERE datname = $1",
      [dbName]
    );

    if (dbCheck.rows.length === 0) {
      console.info(`[Provisioner] Creating database "${dbName}"...`);
      // CREATE DATABASE cannot be executed with parameterized queries, but dbName is sanitized to only allow a-z0-9_
      await pgClient.query(`CREATE DATABASE ${dbName}`);
      console.info(`[Provisioner] Database "${dbName}" created successfully.`);
    } else {
      console.warn(`[Provisioner] Database "${dbName}" already exists.`);
    }
  } finally {
    await pgClient.end();
  }

  // 3. Connect to the newly created database and run migrations
  const tenantDbUrl = `${parsedUrl.protocol}//${parsedUrl.username}:${parsedUrl.password}@${parsedUrl.host}/${dbName}`;
  const tenantPool = new Pool({ connectionString: tenantDbUrl });
  const tenantDb = drizzle(tenantPool);

  try {
    console.info(`[Provisioner] Running migrations on "${dbName}"...`);
    // Resolve absolute path to tenant migrations folder in @sibangku/db
    const migrationsFolder = path.resolve(__dirname, '../../../../db/drizzle/tenant');
    
    await migrate(tenantDb, { migrationsFolder });
    console.info(`[Provisioner] Migrations completed successfully on "${dbName}".`);

    // 4. Seed tenant admin user
    const saltRounds = Number(process.env.BCRYPT_ROUNDS) || 10;
    const passwordHash = await bcrypt.hash(temporaryPassword, saltRounds);

    console.info(`[Provisioner] Seeding admin account (${adminEmail})...`);
    await tenantDb.insert(users).values({
      id: `usr-${tenantSlug}-admin`,
      email: adminEmail,
      name: 'Restaurant Owner',
      passwordHash,
      role: 'TENANT_ADMIN',
      mustChangePassword: true,
      isActive: true,
    });

    // 5. Seed default branding and config settings
    console.info(`[Provisioner] Seeding default configs and branding...`);
    await tenantDb.insert(settings).values([
      {
        key: 'branding',
        value: {
          logo: null,
          favicon: null,
          primaryColor: '#e11d48', // rose-600
          secondaryColor: '#4f46e5', // indigo-600
          font: 'Inter',
          heroImage: null,
          gallery: [],
          socialLinks: {},
          contactInfo: {
            restaurantName,
            email: adminEmail,
          },
        },
      },
      {
        key: 'time_slots',
        value: {
          openingTime: '10:00',
          closingTime: '22:00',
          slotDuration: 30, // in minutes
          reservationDuration: 120, // in minutes
          bufferTime: 15, // in minutes
        },
      },
      {
        key: 'restaurant_profile',
        value: {
          name: restaurantName,
          description: `Selamat datang di ${restaurantName}. Silakan pesan meja dan menu favorit Anda.`,
          address: '',
          phone: '',
          whatsapp: '',
        },
      },
    ]);

    // Seed default categories
    await tenantDb.insert(menuCategories).values([
      { id: 'cat-main', name: 'MAIN COURSE', sortOrder: 1, isActive: true },
      { id: 'cat-drinks', name: 'DRINKS', sortOrder: 2, isActive: true },
      { id: 'cat-dessert', name: 'DESSERT', sortOrder: 3, isActive: true },
    ]);

    console.info(`[Provisioner] Seeding completed for tenant "${dbName}".`);

  } catch (error: any) {
    console.error(`[Provisioner] Error during tenant database configuration:`, error.message);
    throw error;
  } finally {
    await tenantPool.end();
  }

  // 6. Save tenant metadata in control plane database
  const trialStart = new Date();
  const trialEnd = new Date();
  trialEnd.setDate(trialEnd.getDate() + trialDays);

  console.info(`[Provisioner] Saving tenant metadata in control plane...`);
  await db.insert(tenants).values({
    tenantId,
    tenantCode,
    tenantName,
    restaurantName,
    status: 'TRIAL',
    subscriptionStatus: 'TRIAL',
    trialStart,
    trialEnd,
    subscriptionStart: null,
    subscriptionEnd: null,
    databaseIdentifier: dbName,
    webIdentifier,
    apkIdentifier,
    brandingIdentifier,
  });

  // 7. Write audit log entry
  console.info(`[Provisioner] Writing audit logs...`);
  await db.insert(auditLogs).values([
    {
      id: `audit-${tenantId}-created`,
      tenantId,
      action: 'tenant created',
      details: { tenantName, restaurantName, tenantCode },
    },
    {
      id: `audit-${tenantId}-db-created`,
      tenantId,
      action: 'database created',
      details: { dbName },
    },
    {
      id: `audit-${tenantId}-admin-created`,
      tenantId,
      action: 'admin created',
      details: { adminEmail },
    },
    {
      id: `audit-${tenantId}-trial-started`,
      tenantId,
      action: 'trial started',
      details: { trialStart, trialEnd, trialDays },
    },
  ]);

  console.info(`[Provisioner] Tenant ${tenantName} provisioned successfully!`);

  return {
    tenantId,
    tenantCode,
    adminEmail,
    temporaryPassword,
    databaseName: dbName,
    status: 'TRIAL',
  };
}

File: src/SiBangku.ControlApi/Services/TenantProvisioner.cs
using System;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Npgsql;
using SiBangku.Db;
using SiBangku.Shared;
using SiBangku.Shared.Models;
using SiBangku.Shared.Security;

namespace SiBangku.ControlApi.Services
{
    public interface ITenantProvisioner
    {
        Task<ProvisionResult> ProvisionTenantAsync(ProvisionTenantParams paramsDto);
    }

    public class ProvisionTenantParams
    {
        public string TenantName { get; set; } = string.Empty;
        public string RestaurantName { get; set; } = string.Empty;
        public string AdminEmail { get; set; } = string.Empty;
        public string? AdminPassword { get; set; }
        public int TrialDays { get; set; } = 60;
    }

    public class ProvisionResult
    {
        public string TenantId { get; set; } = string.Empty;
        public string TenantCode { get; set; } = string.Empty;
        public string AdminEmail { get; set; } = string.Empty;
        public string TemporaryPassword { get; set; } = string.Empty;
        public string DatabaseName { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
    }

    public class TenantProvisioner : ITenantProvisioner
    {
        private readonly ControlDbContext _controlContext;
        private readonly string _controlDbConnectionString;

        public TenantProvisioner(ControlDbContext controlContext, string controlDbConnectionString)
        {
            _controlContext = controlContext;
            _controlDbConnectionString = controlDbConnectionString;
        }

        public async Task<ProvisionResult> ProvisionTenantAsync(ProvisionTenantParams paramsDto)
        {
            // 1. Generate unique identifiers
            var tenantId = Utils.GenerateTenantId();
            var tenantSlug = Utils.GenerateTenantSlug(paramsDto.TenantName);
            var tenantCode = tenantSlug.ToUpperInvariant();
            var dbName = Utils.GenerateDatabaseIdentifier(tenantSlug);
            var packageId = Utils.GeneratePackageId(tenantSlug);
            var webIdentifier = $"{tenantSlug}.sibangku.example";
            var apkIdentifier = packageId;
            // An operator-supplied password must satisfy the platform password
            // policy; the auto-generated fallback is high-entropy by construction.
            string temporaryPassword;
            if (!string.IsNullOrWhiteSpace(paramsDto.AdminPassword))
            {
                var supplied = paramsDto.AdminPassword.Trim();
                var policyResult = PasswordPolicy.Validate(supplied, paramsDto.AdminEmail);
                if (!policyResult.IsValid)
                {
                    throw new ArgumentException(policyResult.ErrorSummary, nameof(paramsDto));
                }

                temporaryPassword = supplied;
            }
            else
            {
                temporaryPassword = Utils.GenerateTemporaryPassword();
            }

            // provisioning step

            // 2. Create the physical database on PostgreSQL server
            var controlBuilder = new NpgsqlConnectionStringBuilder(_controlDbConnectionString);
            var controlHost = controlBuilder.Host;
            var controlPort = controlBuilder.Port;
            var controlUser = controlBuilder.Username;
            var controlPassword = controlBuilder.Password;

            // Connect to maintenance DB (control DB first, fallback to postgres) to run CREATE DATABASE
            var maintenanceDb = !string.IsNullOrWhiteSpace(controlBuilder.Database) ? controlBuilder.Database : "postgres";
            var systemBuilder = new NpgsqlConnectionStringBuilder
            {
                Host = controlHost,
                Port = controlPort,
                Username = controlUser,
                Password = controlPassword,
                Database = maintenanceDb
            };

            NpgsqlConnection? conn = null;
            try
            {
                conn = new NpgsqlConnection(systemBuilder.ConnectionString);
                await conn.OpenAsync();
            }
            catch
            {
                systemBuilder.Database = "postgres";
                conn = new NpgsqlConnection(systemBuilder.ConnectionString);
                await conn.OpenAsync();
            }

            await using (conn)
            {
                // Check if database exists
                var checkQuery = "SELECT 1 FROM pg_database WHERE datname = @dbName";
                await using (var checkCmd = new NpgsqlCommand(checkQuery, conn))
                {
                    checkCmd.Parameters.AddWithValue("dbName", dbName);
                    var exists = await checkCmd.ExecuteScalarAsync();

                    if (exists == null)
                    {
                        var createQuery = $"CREATE DATABASE \"{dbName}\"";
                        await using (var createCmd = new NpgsqlCommand(createQuery, conn))
                        {
                            await createCmd.ExecuteNonQueryAsync();
                        }
                    }
                }
            }

            // 3. Connect to newly created database and run schema generation
            var tenantBuilder = new NpgsqlConnectionStringBuilder
            {
                Host = controlHost,
                Port = controlPort,
                Username = controlUser,
                Password = controlPassword,
                Database = dbName
            };

            var optionsBuilder = new DbContextOptionsBuilder<TenantDbContext>();
            optionsBuilder.UseNpgsql(tenantBuilder.ConnectionString);

            using (var tenantContext = new TenantDbContext(optionsBuilder.Options))
            {
                // provisioning step
                await tenantContext.Database.EnsureCreatedAsync();
                // provisioning step

                // 4. Seed initial Tenant Admin user
                var passwordHash = PasswordHasher.Hash(temporaryPassword);
                var tenantAdmin = new User
                {
                    UserId = "tenant-admin-init",
                    TenantId = tenantId,
                    Email = paramsDto.AdminEmail,
                    PasswordHash = passwordHash,
                    Name = "Restaurant Owner",
                    Role = "TENANT_ADMIN",
                    MustChangePassword = true,
                    CreatedAt = DateTime.UtcNow
                };

                await tenantContext.Users.AddAsync(tenantAdmin);

                // Seed dynamic setting profiles (PRD §21-22)
                var brandingSetting = new Setting
                {
                    Key = "branding",
                    Value = "{\"primaryColor\":\"#3b82f6\",\"secondaryColor\":\"#1e3a8a\",\"font\":\"Inter\",\"logo\":\"\",\"favicon\":\"\",\"heroImage\":\"\"}"
                };

                var slotsSetting = new Setting
                {
                    Key = "time_slots",
                    Value = "{\"slotDuration\":60,\"maxConcurrentCovers\":30,\"openingTime\":\"08:00\",\"closingTime\":\"22:00\"}"
                };

                // Seed default tables for visual layout
                var defaultTables = new System.Collections.Generic.List<Table>
                {
                    new Table { TableId = $"tbl-1-{tenantId}", TableNumber = "1", Shape = "SQUARE", Capacity = 2, PosX = 80, PosY = 60, Rotation = 0, IsActive = true, CreatedAt = DateTime.UtcNow },
                    new Table { TableId = $"tbl-2-{tenantId}", TableNumber = "2", Shape = "SQUARE", Capacity = 4, PosX = 280, PosY = 60, Rotation = 0, IsActive = true, CreatedAt = DateTime.UtcNow },
                    new Table { TableId = $"tbl-3-{tenantId}", TableNumber = "3", Shape = "ROUND", Capacity = 4, PosX = 80, PosY = 220, Rotation = 0, IsActive = true, CreatedAt = DateTime.UtcNow },
                    new Table { TableId = $"tbl-4-{tenantId}", TableNumber = "4", Shape = "RECTANGLE", Capacity = 6, PosX = 280, PosY = 220, Rotation = 0, IsActive = true, CreatedAt = DateTime.UtcNow }
                };

                await tenantContext.Settings.AddRangeAsync(brandingSetting, slotsSetting);
                await tenantContext.Tables.AddRangeAsync(defaultTables);
                await tenantContext.SaveChangesAsync();
                // provisioning step
            }

            // 5. Save tenant record and log audit details in Control Plane
            var now = DateTime.UtcNow;
            var tenant = new Tenant
            {
                TenantId = tenantId,
                TenantCode = tenantCode,
                TenantName = paramsDto.TenantName,
                RestaurantName = paramsDto.RestaurantName,
                Status = "TRIAL",
                SubscriptionStatus = "TRIAL",
                TrialStart = now,
                TrialEnd = now.AddDays(paramsDto.TrialDays),
                DatabaseIdentifier = dbName,
                StorageIdentifier = $"storage_{tenantSlug}",
                WebIdentifier = webIdentifier,
                ApkIdentifier = apkIdentifier,
                CreatedAt = now,
                UpdatedAt = now
            };

            await _controlContext.Tenants.AddAsync(tenant);

            var audit = new AuditLog
            {
                Id = $"aud-{Guid.NewGuid():N}",
                TenantId = tenantId,
                Action = "provision tenant",
                UserId = "system-api",
                Details = $"{{\"provisionedAt\":\"{now:O}\",\"database\":\"{dbName}\",\"admin\":\"{paramsDto.AdminEmail}\",\"trialDays\":{paramsDto.TrialDays}}}",
                CreatedAt = now
            };

            await _controlContext.AuditLogs.AddAsync(audit);
            await _controlContext.SaveChangesAsync();

            // Automatically scaffold tenant workspace files (desktop bat, web launcher, android project, config)
            TenantWorkspaceScaffolder.ScaffoldWorkspace(
                tenantCode,
                paramsDto.RestaurantName,
                tenantId,
                paramsDto.TenantName,
                dbName);

            return new ProvisionResult
            {
                TenantId = tenantId,
                TenantCode = tenantCode,
                AdminEmail = paramsDto.AdminEmail,
                TemporaryPassword = temporaryPassword,
                DatabaseName = dbName,
                Status = "TRIAL"
            };
        }
    }
}
